### DISCUSSION TITLE
About of payment gateway development

### DISCUSSION THREAD

**Discussion by 0xBeycan** [impress-org/givewp#7304] [Developer Q&A]
We developed an integration for our CryptoPay plugin. It works fine in Visual Forms, but unfortunately it does not redirect in the Legacy system. Can you help with this issue?

Here is repo, you can check it.

https://github.com/BeycanPress/givewp-cryptopay-gateway

**Comment by glaubersilva** [comment]
Hi @BeycanDeveloper 

Hope you doing well. 

We need better documentation for it, but for now, if you take a look at [this class](https://github.com/impress-org/givewp/blob/develop/src/DonationForms/Listeners/AddRedirectUrlsToGatewayData.php#L22-L23) you'll see that we provide in the `$gatewayData` var some default redirect URLs that work for both: option-based (legacy) and visual builder forms. 

To use it in your integration, all you need to do is replace this line:

`'givewp-success-url' => urlencode(give_get_success_page_uri())`

By this one: 

`'givewp-success-url' => $gatewayData['successUrl'],`

Let me know if it works for you!

**Comment by 0xBeycan** [comment]
In fact, when it succeeds, there is no problem with routing. The "createPayment" method returns the link to the payment page created by our plugin with the "RedirectOffsite" class. However, Legacy Form tries to open it on the same page instead of redirecting to this page.

```php
    public function createPayment(Donation $donation, $gatewayData): RedirectOffsite
    {
        return Processor::createPayment(Type::LITE, $donation, $this->generateSecureGatewayRouteUrl(
            'handleCreatePaymentRedirect',
            $donation->id,
            [
                'givewp-donation-id' => $donation->id,
                'givewp-payment-token' => 'cp_payment_token',
                'givewp-transaction-id' => 'cp_transaction_id',
                'givewp-success-url' => urlencode(give_get_success_page_uri()),
            ]
        ));
    }
```

https://github.com/BeycanPress/givewp-cryptopay-gateway/blob/c3e12baba1eabbc38810729a58a2ff26d0545e76/app/Gateways/GatewayPro.php#L84

https://github.com/BeycanPress/givewp-cryptopay-gateway/blob/c3e12baba1eabbc38810729a58a2ff26d0545e76/app/Gateways/Processor.php#L91

**Comment by glaubersilva** [comment]
Hi @BeycanDeveloper 

Not sure if I understand what you said. So I would like to confirm a few things:

1. Are you having the same problem even using the `$gatewayData['successUrl']` value?
2. For legacy forms, is the `createPayment()` method not redirecting to the external payment page?

If the reply to the last question is yes, let me know what happens when you replace your `createPayment()` method with this one:

```
public function createPayment(Donation $donation, $gatewayData): RedirectOffsite
{
     return new RedirectOffsite('https://givewp.com/');        
}
```

**Comment by 0xBeycan** [comment]
Yes second one and already i am using RedirectOffsite class.

![image](https://github.com/impress-org/givewp/assets/50718965/d7c8b14b-36fa-47cb-8fbd-2178c5484af6)

**Reply by glaubersilva** [reply]
@BeycanDeveloper I know you are already using the `RedirectOffsite` class, but I was trying to check what happens when you replace the link generated by the `Helpers::createSPP()` method with any other link.

In my tests, when I replace the return value of the `createPayment()` method with the `https://givewp.com/` URL, the redirect works as expected. 

I just noticed that the `Helpers::createSPP()` method is returning this: `return home_url("/?cp_spp={$token}");`

So, seems it's using [the home_url() function from the WP core](https://developer.wordpress.org/reference/functions/home_url/) and it should explain why the redirect to the external page isn't working.

**Comment by 0xBeycan** [comment]
So with this logic, it expects an external link. If it's the same site, it tries to open it inside. So how do we prevent this? Because our plugin works completely within WordPress.

**Reply by glaubersilva** [reply]
@BeycanDeveloper Sorry for the late reply here.

Yes, you are right. The `RedirectOffsite` wasn't designed to be used with internal URLs from the WordPress website.

If your gateway doesn't provide an external page, ideally, loading the gateway fields on the donation form is the most indicated.

However, there is a workaround to allow you to simulate an off-site gateway using an on-site link, so I'll provide a few samples for these two approaches.

**Approach 1 - Loading the gateway fields on the donation form.**

For this approach, you should handle the transaction creation inside the `createPayment()` method and return another `GatewayCommand` different from the `RedirectOffsite`. 

A few options would be: 

```
return new PaymentProcessing($transactionId);
return new PaymentComplete($transactionId);
```

Also, to use this approach on _Legacy Forms_ is necessary to implement the `getLegacyFormFieldMarkup()` method in the gateway class, here is a sample: https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php#L105-L108

Here is the CreditCardForm trait used in the sample above:
https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/Traits/CreditCardForm.php

The final result is something like this:

![image](https://github.com/impress-org/givewp/assets/16308864/b9104bfa-5232-4b47-9a38-9dd32d5028d6)


For the new _Visual Form Builder Forms_, you need to implement the gateway frontend logic in a separate JS/tsx file, here is a sample: https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/StripePaymentElementGateway/stripePaymentElementGateway.tsx

Then, you need to enqueue this script in the gateway class through the `enqueueScript()` method. You also can pass any value from the backend to the frontend file through the `formSettings()` method. 

Here is a sample using both methods: https://github.com/impress-org/givewp/blob/develop/src/PaymentGateways/Gateways/Stripe/StripePaymentElementGateway/StripePaymentElementGateway.php#L54-L89

The final result is something like this:

![image](https://github.com/impress-org/givewp/assets/16308864/bd697249-c7ff-43d6-bbba-fc1f2c94f1c0)

**Approach 2 - Simulating a redirect off-site gateway.**

Let's suppose that you are redirecting the users to the following URL:

```
return new RedirectOffsite(add_query_arg('off-site-gateway-simulation', true, home_url())); //https://your-site-home-url.com/?off-site-gateway-simulation=true
```

So, to simulate an off-site gateway, you need to use the WordPress `init` hook  to intercept the original redirect and replace it with something like this:

```
add_action('init', function () {
    if ( ! isset($_GET['off-site-gateway-simulation'])) {
        return;
    }

    if (isLegacyFormReferrer()) {
        echo '<script>window.top.location.href ="' . home_url($_SERVER['REQUEST_URI']) . '";</script>';
        exit();
    }

    echo 'Off-site Gateway Simulation - You should load your "off-site" gateway here...';

    exit();
});

function isLegacyFormReferrer(): bool
{
    //V2 referrer: https://example.com/give/v2-tests?giveDonationFormInIframe=1
    //V3 referrer: https://example.com/?givewp-route=donation-form-view&form-id=1350
    return strpos($_SERVER['HTTP_REFERER'], 'giveDonationFormInIframe') !== false;
}
```
